src/tools/nx.8
.TH nx 8 User Manuals
.SH NAME
nx \- Network Context Setup
.SH SYNOPSIS
\fBnx -create \fInid\fB [\fIlist\fB] [-- \fIprogram\fB \fIargs\fB*]

nx -migrate \fInid\fB -- \fIprogram\fB \fIargs\fB*

nx -set-flags \fInid\fB \fIlist\fB

nx -get-flags \fInid\fB

nx -add-addr \fInid\fB \fIaddr\fB/\fInetmask\fB*

nx -rem-addr \fInid\fB \fIaddr\fB/\fInetmask\fB*
\f1
.SH DESCRIPTION
While the Context Separation is sufficient to isolate groups of processes, a different kind of separation, or rather a limitation, is required to confine processes to a subset of available network addresses.

Several issues have to be considered when doing so; for example, the fact that bindings to special addresses like IPADDR_ANY or the local host address have to be handled in a very special way.

Currently, Linux-VServer doesn't make use of virtual network devices (and maybe never will) to minimize the resulting overhead. Therefore socket binding and packet transmission have been adjusted.
.SH OPTIONS
.TP
\fB-create\f1
Create a new network context with id \fInid\f1 using optional flags in \fIlist\f1 and optionally run \fIprogram\f1 in the new context.
.TP
\fB-migrate\f1
Migrate to network context \fInid\f1 and execute \fIprogram\f1.
.TP
\fB-set-flags\f1
Set context flags in \fIlist\f1 for network context \fInid\f1. See below for valid flags.
.TP
\fB-get-flags\f1
Get flags for network context \fInid\f1. See below for valid flags.
.TP
\fB-add-addr\f1
Add IP address \fIaddr\f1/\fInetmask\f1 to network context \fInid\f1. More than one IP address can be specified.
.TP
\fB-rem-addr\f1
Remove IP address \fIaddr\f1/\fInetmask\f1 from network context \fInid\f1. More than one IP address can be specified.
.SH VALID FLAGS
STATE_SETUP - network context is in setup state

SC_HELPER - network state change helper

PERSISTENT - make network context persistent
.SH AUTHOR
nx was written by Benedikt Boehm <hollow@gentoo.org>

You will find the newest version of dx in the libvserver package at \fBhttp://linux-vserver.org\f1. There is also a mailing list, you will find information about it at Linux-VServer main site.
.SH BUGS
You should report them to the mailing list.
.SH SEE ALSO
\fBdx(8)\f1, \fBix(8)\f1, \fBns(8)\f1, \fBvr(8)\f1, \fBvx(8)\f1, 


NEWS
01 Aug 2005: First stable release of libvserver

I'm pleased to announce libvserver 0.3, the first release considered more or
less stable. This release supports the complete API of the Linux-VServer 2.0
kernel patches (libvserver.so) and provides additional helper functions such as
capability lists (libvserver-util.so). As of now, the public interface to
libvserver.so is not subject of change anymore.

An API doc is also available via doxygen, just use the provided Doxyfile or go
to http://dev.gentoo.org/~hollow/vserver/libvserver/doc/ for the latest API doc.

Have fun!


src/tools/nx.c
// Copyright 2006 Benedikt Böhm <hollow@gentoo.org>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the
// Free Software Foundation, Inc.,
// 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <arpa/inet.h>

#include "tools.h"
#include "vserver.h"

static const char *rcsid = "$Id: nx.c 294 2006-07-09 08:46:15Z hollow $";

static
struct option long_opts[] = {
	COMMON_LONG_OPTS
	{ "create",    1, 0, 0x10 },
	{ "migrate",   1, 0, 0x11 },
	{ "set-flags", 1, 0, 0x12 },
	{ "get-flags", 1, 0, 0x13 },
	{ "add-addr",  1, 0, 0x14 },
	{ "rem-addr",  1, 0, 0x15 },
	{ NULL,        0, 0, 0 },
};

static inline
void usage(int rc)
{
	printf("Usage:\n\n"
	          "nx -create    <nid> [<list>] [-- <program> <args>*]\n"
	          "   -migrate   <nid> -- <program> <args>*\n"
	          "   -set-flags <nid> <list>\n"
	          "   -get-flags <nid>\n"
	          "   -add-addr  <nid> <addr>/<netmask>*\n"
	          "   -rem-addr  <nid> <addr>/<netmask>*\n");
	exit(rc);
}

static
int str_to_addr(char *str, uint32_t *ip, uint32_t *mask)
{
	struct in_addr ib;
	char *addr_ip, *addr_mask;
	
	*ip   = 0;
	*mask = 0;
	
	addr_ip   = strtok(str, "/");
	addr_mask = strtok(NULL, "/");
	
	if (addr_ip == 0)
		return -1;
	
	if (inet_aton(addr_ip, &ib) == -1)
		return -1;
	
	*ip = ib.s_addr;
	
	if (addr_mask == 0) {
		/* default to /24 */
		*mask = ntohl(0xffffff00);
	} else {
		if (strchr(addr_mask, '.') == 0) {
			/* We have CIDR notation */
			int sz = atoi(addr_mask);
			
			for (*mask = 0; sz > 0; --sz) {
				*mask >>= 1;
				*mask  |= 0x80000000;
			}
			
			*mask = ntohl(*mask);
		} else {
			/* Standard netmask notation */
			if (inet_aton(addr_mask, &ib) == -1)
				return -1;
			
			*mask = ib.s_addr;
		}
	}
	
	return 0;
}

int main(int argc, char *argv[])
{
	INIT_ARGV0
	
	int c, i;
	nid_t nid = 0;
	char *buf;
	
	/* syscall data */
	struct nx_create_flags cf = {
		.flags = 0,
	};
	
	struct nx_flags flags = {
		.flags = 0,
		.mask  = 0,
	};
	
	struct nx_addr addr;
	
	uint64_t mask;
	
#define CASE_GOTO(ID, P) case ID: nid = atoi(optarg); goto P; break
	
	/* parse command line */
	while (GETOPT(c)) {
		switch (c) {
			COMMON_GETOPT_CASES
			
			CASE_GOTO(0x10, create);
			CASE_GOTO(0x11, migrate);
			CASE_GOTO(0x12, setflags);
			CASE_GOTO(0x13, getflags);
			CASE_GOTO(0x14, addaddr);
			CASE_GOTO(0x15, remaddr);
			
			DEFAULT_GETOPT_CASES
		}
	}
	
#undef CASE_GOTO
	
	goto usage;
	
create:
	if (argc > optind && strcmp(argv[optind], "--") != 0)
		if (flist64_from_str(argv[optind], nflags_list, &cf.flags, &mask, '~', ',') == -1)
			perr("flist64_from_str");
	
	if (nx_create(nid, &cf) == -1)
		perr("nx_create");
	
	if (argc > optind+1)
		execvp(argv[optind+1], argv+optind+1);
	
	goto out;

migrate:
	if (nx_migrate(nid) == -1)
		perr("nx_migrate");
	
	if (argc > optind+1)
		execvp(argv[optind+1], argv+optind+1);
	
	goto out;
	
setflags:
	if (argc <= optind)
		goto usage;
	
	if (flist64_from_str(argv[optind], nflags_list, &flags.flags, &flags.mask, '~', ',') == -1)
		perr("flist64_from_str");
	
	if (nx_set_flags(nid, &flags) == -1)
		perr("nx_set_flags");
	
	goto out;
	
getflags:
	if (nx_get_flags(nid, &flags) == -1)
		perr("nx_get_flags");
	
	if (!(buf = flist64_to_str(nflags_list, flags.flags, '\n')))
		perr("flist64_to_str");
	
	printf("%s", buf);
	free(buf);
	
	goto out;
	
addaddr:
	if (argc <= optind)
		goto usage;
	
	for (i = optind; argc > i; i++) {
		addr.type  = NXA_TYPE_IPV4;
		addr.count = 1;
		
		if (str_to_addr(argv[i], &addr.ip[0], &addr.mask[0]) == -1)
			perr("str_to_addr");
		
		if (nx_add_addr(nid, &addr) == -1)
			perr("nx_add_addr");
	}
	
	goto out;
	
remaddr:
	if (argc <= optind)
		goto usage;
	
	for (i = optind; argc > i; i++) {
		addr.type  = NXA_TYPE_IPV4;
		addr.count = 1;
		
		if (str_to_addr(argv[i], &addr.ip[0], &addr.mask[0]) == -1)
			perr("str_to_addr");
		
		if (nx_rem_addr(nid, &addr) == -1)
			perr("nx_rem_addr");
	}
	
	goto out;
	
usage:
	usage(EXIT_FAILURE);

out:
	exit(EXIT_SUCCESS);
}
